src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -284,90 +284,109 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
+                // Related document must come from a distinct project.
+                Contract.ThrowIfTrue(relatedDocumentId.ProjectId == addedDocumentId.ProjectId);
+
+                relatedProjectIdHint = relatedDocumentId.ProjectId;
+                relatedDocumentIdsAndStates.Add((addedDocumentId, relatedDocument.DocumentState));
+            }
 
-            // Related document must come from a distinct project.
-            Contract.ThrowIfTrue(relatedDocumentId.ProjectId == addedDocumentId.ProjectId);
+            if (relatedDocumentIdsAndStates.IsEmpty)
+                return solution;
 
-            var newSolution = solution.WithDocumentContentsFrom(addedDocumentId, relatedDocument.DocumentState, forceEvenIfTreesWouldDiffer: false);
-            return (newSolution, relatedProjectId: relatedDocumentId.ProjectId);
+            return solution.WithDocumentContentsFrom(relatedDocumentIdsAndStates.ToImmutableAndClear(), forceEvenIfTreesWouldDiffer: false);
         }
 
-        static Solution UpdateExistingDocumentsToChangedDocumentContents(Solution solution, DocumentId changedDocumentId, HashSet<DocumentId> processedDocuments)
+        static Solution UpdateExistingDocumentsToChangedDocumentContents(Solution solution, HashSet<DocumentId> changedDocumentIds)
         {
             // Changing a document in a linked-doc-chain will end up producing N changed documents.  We only want to
             // process that chain once.
-            if (processedDocuments.Add(changedDocumentId))
+            using var _ = PooledDictionary<DocumentId, DocumentState>.GetInstance(out var documentIdsToUpdate);
+
+            foreach (var changedDocumentId in changedDocumentIds)
             {
-                var changedDocument = solution.GetRequiredDocument(changedDocumentId);
+                Document? changedDocument = null;
                 var relatedDocumentIds = solution.GetRelatedDocumentIds(changedDocumentId);
+
                 foreach (var relatedDocumentId in relatedDocumentIds)
                 {
                     if (relatedDocumentId == changedDocumentId)
                         continue;
 
-                    if (processedDocuments.Add(relatedDocumentId))
-                        solution = solution.WithDocumentContentsFrom(relatedDocumentId, changedDocument.DocumentState, forceEvenIfTreesWouldDiffer: false);
+                    if (!changedDocumentIds.Contains(relatedDocumentId))
+                    {
+                        changedDocument ??= solution.GetRequiredDocument(changedDocumentId);
+                        documentIdsToUpdate[relatedDocumentId] = changedDocument.DocumentState;
+                    }
                 }
             }
 
-            return solution;
+            if (documentIdsToUpdate.Count == 0)
+                return solution;
+
+            var toProcessRelatedDocumentIds = documentIdsToUpdate.SelectAsArray(kvp => (kvp.Key, kvp.Value));
```

Comment:
i can't quite reason about everything here.  do we really ever need a map?  couldn't an array work for this?  

also, toProcessRelatedDocumentIds  seems like the wrong name, since it's more than just the id.

Reply:
As I mentioned in the other comment, it seems like it could need a map. I'll change the name.